SNIPS/snip-x.md

```diff
@@ -0,0 +1,101 @@
+token transfers, allowing it to be broadly compatible with Starknet DeFi, and also requires no changes to the token
+interface. This is possible because of Starknet's native account abstraction and multicall.
+
+## Specification
+
+```cairo
+#[starknet::interface]
+pub trait IFeeOnTransferToken<TContractState> {
+    // Gets the amount of fees already paid for the given sender
+    fn get_fees_paid(self: @TContractState, sender: ContractAddress) -> ContractAddress;
+
+    // Returns the amount of fees required to transfer the specified amount of tokens from the given sender to the receiver.
+    fn compute_fees_required(
+        self: @TContractState, sender: ContractAddress, receiver: ContractAddress, amount: u128
+    ) -> u128;
+
+    // Pay fees from the given address for the specified sender.
+    // If `from` is the caller, then it pays from the caller's balance.
+    // Otherwise, it pays from the allowance of the `from` address to the caller as the spender.
+    // Returns the total amount of fees paid for the sender
+    fn pay_fees_verbose(
+        ref self: TContractState, from: ContractAddress, sender: ContractAddress, amount: u128
+    ) -> u128;
+    
+    // Same as pay_fees_verbose but always pays `from` the caller address
+    fn pay_fees_for_sender(ref self: TContractState, sender: ContractAddress, amount: u128) -> u128;
+    
+    // Same as pay_fees_for_sender, but the sender is always the caller address
+    fn pay_fees(ref self: TContractState, amount: u128) -> u128;
+
+    // Withdraws any fees paid for the given sender to the specified recipient address.
+    // This can be called  by anyone for any address--fees are a transient payment location to enable transfers for a given address. Leftover fees should always be withdrawn in the same transaction.
+    // Returns the amount of fees that were withdrawn.
+    fn withdraw_fees_paid_verbose(
+        ref self: TContractState, sender: ContractAddress, recipient: ContractAddress
+    ) -> u128;
+    
+    // Same as `withdraw_fees_paid_verbose` but recipient is always the caller
+    fn withdraw_fees_paid_for_sender(ref self: TContractState, sender: ContractAddress) -> u128;
+    
+    // Same as `withdraw_fees_paid_for_sender` but the sender and recipient are both set to the caller
+    fn withdraw_fees_paid(ref self: TContractState) -> u128;
+}
+```
+
+The expected usage flow is as follows:
+
+- DApp stores off-chain metadata about whether a token is fee-on-transfer, and implements the following logic if it is
+- When depositing this token into a dapp, the `recipient` is the Dapp contract:
+    - Dapp calls `compute_fees_required(user, dapp_contract, amount)` off-chain
+    - Dapp adds `pay_fees(computed_fees)` call to the list of calls before all dapp calls
+    - Dapp adds rest of calls as normal
+    - Dapp adds `withdraw_fees_paid()` call to return any overpaid transfer fees
```

Comment:
shouldn't this be `withdraw_fees_paid_for_sender(user)` ?

Reply:
again this is from the user account's perspective